### ISSUE TITLE
[Auth] Inner exception #0

### ISSUE THREAD

**Issue by AndresCuicui** [firebase/quickstart-unity#866]
### [REQUIRED] Please fill in the following fields:

  * Unity editor version: 2019.4.8f1
  * Firebase Unity SDK version: 6.15.2
  * Source you installed the SDK: .unitypackage
  * Problematic Firebase Component: Auth
  * Other Firebase Components in use: Analytics, FIrestore
  * Additional SDKs you are using: 
  * Platform you are using the Unity editor on: Windows
  * Platform you are targeting: iOS, Android
  * Scripting Runtime: Mono

### [REQUIRED] Please describe the issue here:

We keep getting this exception when trying to authenticate by using anonymous, mail and password or Google Play Games authentication:

```
SignInAnonymouslyAsync encountered an error: System.AggregateException: One or more errors occurred. ---> System.AggregateException: One or more errors occurred. ---> Firebase.FirebaseException: An internal error has occurred.
   --- End of inner exception stack trace ---
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.AggregateException: One or more errors occurred. ---> Firebase.FirebaseException: An internal error has occurred.
   --- End of inner exception stack trace ---
---> (Inner Exception #0) Firebase.FirebaseException: An internal error has occurred.<---
<---
```
It is completely impossible to authenticate.
[Edit] Tried to connect to Firebase project from an empty Unity Project. Keep getting this exception. Seems like project inside Firebase got messed up.

#### Steps to reproduce:

We first implemented the email + password sign in. Everything worked fine.
Then the anonymous sign in (and the link afther that). Everything worked fine.
Then added the Google Play Games authentication, which was hard to implement, but managed to make it work. Everything worked fine.
Following day: we get the exception. No code changes or anything.

What happened? How can we make the problem occur?

We simply don't know what happened. One day everything was working perfectly, the following day we were unable to authenticate. Furthermore, we went back some commits and the Unity project is completely ruined (even to the first commits where we only had the mail + password sign in method). There's no way to make it authenticate. What's the Inner exception #0? How can we solve it? We've seen a lot of issues and posts on the internet (stack overflow and everything) and there's no solution. And also the docs are not helpful.

#### Relevant Code:

[Edit] Here is the example auth code we tested on an empty Unity project trying to connect to our Firebase project. It's literally the only code in the whole project:

```
using Firebase;
using Firebase.Auth;
using System.Collections;
using System.Collections.Generic;
using TMPro;
using UnityEngine;

public class FirebaseManager : MonoBehaviour
{

    public TMP_InputField mail;
    public TMP_InputField pass;
    public TMP_Text output;

    private FirebaseAuth auth;
    private FirebaseUser user;
    private FirebaseApp app;

    private bool firebaseAvailable;

    private void Awake()
    {
        firebaseAvailable = false;
    }

    private void AuthStateChanged(object sender, System.EventArgs e)
    {
        if(auth.CurrentUser != user)
        {
            bool signedIn = user != auth.CurrentUser && auth.CurrentUser != null;
            if (!signedIn && user != null)
            {
                string msg = "Signed out " + user.UserId + "\n";
                Debug.LogError(msg);
                if (output != null)
                {
                    output.text += msg;
                }
            }
            user = auth.CurrentUser;
            if (signedIn)
            {
                string msg = "Signed in " + user.UserId + "\n";
                Debug.Log(msg);
                if (output != null)
                {
                    output.text += msg;
                    output.text += "user.DisplayName = " + user.DisplayName + "\n";
                    output.text += "user.Email = " + user.Email + "\n";
                    output.text += "user.PhotoUrl = " + user.PhotoUrl + "\n";
                }

            }
        }
    }

    // Start is called before the first frame update
    void Start()
    {
        FirebaseApp.CheckAndFixDependenciesAsync().ContinueWith(task => {
            var dependencyStatus = task.Result;
            if (dependencyStatus == DependencyStatus.Available)
            {
                firebaseAvailable = true;
                app = FirebaseApp.DefaultInstance;
                auth = FirebaseAuth.DefaultInstance;
                auth.StateChanged += AuthStateChanged;
                AuthStateChanged(this, null);

            }
            else
            {
                Debug.LogError("Unable to initialize Firebase");
            }
        });
    }

    public void SignInMailPass()
    {
        if (!firebaseAvailable)
        {
            Debug.LogError("Firebase not available.");
            return;
        }
        auth.SignInWithEmailAndPasswordAsync(mail.text, pass.text).ContinueWith(task => {
            if (task.IsCanceled)
            {
                Debug.LogError("SignInWithEmailAndPasswordAsync was canceled.");
                return;
            }
            if (task.IsFaulted)
            {
                Debug.LogError("SignInWithEmailAndPasswordAsync encountered an error: " + task.Exception);
                if (output != null)
                {
                    output.text += task.Exception;
                }
                return;
            }
            user = task.Result;
            string msg = "User signed in successfully: " + user.DisplayName + ", " + user.UserId + "\n";
            Debug.Log(msg);
            if (output != null)
            {
                output.text += msg;
            }
        });
    }
}
```
[edit] updated the code.

**Comment by AndresCuicui** [comment]
We tried connecting to our Firebase app from an empty Unity project and got same result. Added the code to the issue.

**Comment by cynthiajoan** [comment]
Hi @AndresCuicui, can you try add this block of code in start()
`public virtual void Start() {
      Firebase.FirebaseApp.CheckAndFixDependenciesAsync().ContinueWithOnMainThread(task => {
        dependencyStatus = task.Result;
        if (dependencyStatus == Firebase.DependencyStatus.Available) {
          Awake();
        } else {
          Debug.LogError(
            "Could not resolve all Firebase dependencies: " + dependencyStatus);
        }
      });
    }`

We usually want to make sure all the dependencies are resolved to make sure Firebase App is initialized.

Please take this [file](https://github.com/firebase/quickstart-unity/blob/master/auth/testapp/Assets/Firebase/Sample/Auth/UIHandler.cs) for reference.

**Comment by AndresCuicui** [comment]
Hi @cynthiajoan, yes. As soon I posted the code I realised it was missing the checkandfixdependencies and added it on my side (should have updated the post). Thanks.
Output remains the same.

**Comment by chkuang-g** [comment]
Hi @AndresCuicui 

I would recommend you to use `ContinueWithOnMainThread()` instead of `ContinueWith()` for two reasons
1. When using `ContinueWith()`, the continuation function usually is called from a non-main thread.  And Unity API in general don't like that.
2. If any exception is thrown from a non-main thread, some version of Unity would just crash silently.

Second, can you try to catch if any exception is thrown in your continuation function?

Ex.

```
        FirebaseApp.CheckAndFixDependenciesAsync().ContinueWith(task => {
            try{
            var dependencyStatus = task.Result;
            if (dependencyStatus == DependencyStatus.Available)
            {
                firebaseAvailable = true;
                app = FirebaseApp.DefaultInstance;
                auth = FirebaseAuth.DefaultInstance;
                auth.StateChanged += AuthStateChanged;
                AuthStateChanged(this, null);

            }
            else
            {
                Debug.LogError("Unable to initialize Firebase");
            }
            } catch (System.Exception e) {
               Debug.LogException(e);
            }
        });
```

**Comment by AndresCuicui** [comment]
Hi @chkuang-g, thanks for the reply.
The code I posted was from a test with an empty Unity project and using only the code shown in the docs, just to make sure It was not a problem with the code. In our project, we do use the ContinueWithOnMainThread() function since we are aware that Unity requires some functions to be executed in the main thread. But thanks for the suggestion.
Also, we'll try using a try-catch statement within the continuation function.

As an update on this thread:

We got suggested by Firebase support to try an empty project (both Unity project and Firebase project), which we did. We found out that everything worked smoothly, so we tried connecting the new Unity project to our old Firebase project and found out that the code that worked for the new Firebase project did not work for the old Firebase project. So we run the four possible situations:

New Unity project + new Firebase project => Working.
New Unity project + old Firebase project => Not working.
Old Unity project + new Firebase project => Working.
Old Unity project + old Firebase project => Not working.

From those tests, it's easy to see that the code of our Unity project is not the problem. So, what's the difference between our old Firebase project and the new one? The old one has **Google Play Games** authentication system enabled. And here's where everything gets messy. I'll explain:
We first added the mail+pass authentication system to our game (we need to keep track of user data, inventory and stuff). Everything worked perfectly.
We then added the anonymous authentication system (plus the linking and everything) and everything kept working smoothly.
Here comes the messy part: We then added the Google Play Games authentication system.
As you guys know, it's a task that require some configuration, but we managed to make it work on the Firebase side. We even got test accounts registered using the Play Games tokens and everything. It _seemed_ fine. BUT. As a part of the Google Play Games API, apps are required to show an **oauth consent screen**. Since our app is a new one (not our first app, but the first one in which we use Play Games API), we had to create the app in the google console and register everything. But, **upon requesting verification, we got rejected** (for some silly things like the icon and the website) **and then everything broke**.

**It seems like the failure in the Google verification process broke everything inside the Firebase project**. After that, it was completely impossible to use the Firebase API. We tried removing the Google Play Games authentication system from our Firebase Project and even from the project's code but nothing worked. The Firebase project is completely broken.

[A little off-topic] And here it gets even worse: Since the old Firebase project was broken, but we still wanted to use Firebase for our game (because of its useful features), we deleted our old Firebase project and created a new one. That was last friday. Now on monday, 3 days later, I login to our Google Console to find out that our app is being deleted without being requested by us. **It seems like deleting your Google Play Games linked Firebase project also tries to delete your Google Play app!!!! It's crazy!!** How can that happen? I know that Firebase is owned and operated by Google, but how on earth can deleting a service also delete the app?

So, I thought about closing the issue, but this really is too important. We do need to know how to get the info on those Inner exception #0. We (developers) seriously need to.
